### ISSUE TITLE
Server stopped working in October 2022?

### ISSUE THREAD

**Issue by MartinvP** [robots/4dayforecast#8]
It seems we have a new issue with the 4 Day Forecast program. As of Sat. Oct 22nd 2022 there is no more forecast information on my console. Has anyone still got one of these old stations and does anyone have a clue as to which server could provide forecast data?

**Comment by SedonD** [comment]
Same here, this stuff served me well for a long time.
Maybe it's time to look for a new one, but it's hard to find a replacement for such this complex unit... ):

**Comment by mati159** [comment]
I also have problems with it but yesterday i've managed to get update after 6 days or so. It costed me 2 banned API keys and still it's not fixed. The problem I still have is that the program can get into update loop mode where it constantly tries to update weather to the point of using all free API calls. I don't know when (wayback machine says after August), but weatherbit changed terms of free service so now we have 50 calls a day instead of 500, 1 call a second instead of 5 and 7 days instead of 16. I suspect two issues: (I'm not a programmer to be clear)
1. getweather makes two calls, one for forecast and one for current weather, if they are made too fast it could make two calls a second and get an error/corrupted data
2. I don't know how this program handles non-complete data but I suspect that it can expect to get data for 10 days instead of 7 and maybe tries to download more and gets into loop (I had about 400 calls from 0:00 to 15:00)

I may try to create non-professional solution to both problems by making delay in data calls and "create" days 8-10 by copying days 1-3 but on Saturday or Sunday and I have to find a way to force the script to read data from local file instead of server because of 50 calls limit

I hope someone will find a decent solution for us

EDIT1:

If someone is interested there are more issues I had in past month:
1. Corrupted savefile? - NET framework exeption - Index outside array - it was about month ago and I reinstalled program and created new savefile (it was made in second try, more explanation in next issue
2. I wasn't able to create new savefile yesterday when I tried to fix ix (partially succesfully). After choosing my city and clicking save, home location wasn't saved and was still blank. I tried many times to reinstall different versions of 4-day Forecast, installing NET Framework update, VC++ updates and still nothing. Only working solution for me was finding older data folder in trash (I've deleted it because i thought it was corrupted) and after that (and new API key) I got the update
3. After updating to Windows 10 I have to replug transmitter, because it doesn't work without it. I have it plugged to my regular PC so I turn  it on and off daily and every time I have to do this

EDIT2:

Maybe someone can revive code which was downloading data from openweathermaps if there is no other option. I don't remember using it so I don't know why it was changed but it is probably easier than changing to completely new provider

**Comment by gino0631** [comment]
Same here. As `getweather.php` caches forecast but not current weather responses, it's not surprising the new limit of 50 calls was hit. I'd fix that by caching current weather responses as well, say, for 1 hour.

**Comment by cmumathwhiz** [comment]
I was also affected by the issue here - and thanks to the tips posted here - I was able to get it to work.  I'm not very good at pull requests and updating code, but I'm more than happy to share what I did.  In short - I changed the URL to do days=7 instead of days=10 (that stopped errors from getting the data from weatherbit), then in the loops I only looped through 7 (that cleared up the PHP errors from trying to access indexes that didn't exist), and created another loop for 8-10 to duplicate day 7 into those days (to allow the forecast program to get the same amount of data it is expecting). Here is my after weather.php code (with my API removed) that is working for me - at least as of now - only just started running it again - will check to see if I hit the API access limits again.
[getweather.txt](https://github.com/robots/4dayforecast/files/9833332/getweather.txt)

**Comment by gino0631** [comment]
@cmumathwhiz as the results of current weather responses are still not cached, it should stop working once called more than 50 times a day.

**Comment by cmumathwhiz** [comment]
@gino0631 Agreed - but I don't leave my weather app up on my computer - only use it to transmit to my device.  I checked my webserver logs and it doesn't call to get new weather data all that often.  I'll keep monitoring to see if I run into any issue. If I run into issues - I'll look at adding cache for the current data.

**Comment by gino0631** [comment]
@cmumathwhiz OK. I'll also take a look, but only tomorrow (I hope they unblock me by that time).

**Comment by gino0631** [comment]
My implementation of the corrections is here: gino0631/4dayforecast@b66b593dca3bea96e7c220cbb12d900a966ed06c

**Comment by robots** [comment]
merged, thanks for fix :-)
